ext/src/http/client/curl/http_client_curl.cc

```diff
@@ -474,13 +474,20 @@ bool HttpClient::MaybeSpawnBackgroundThread()
               CURLcode result   = msg->data.result;
               Session *session  = nullptr;
               curl_easy_getinfo(easy_handle, CURLINFO_PRIVATE, &session);
-              // If it's already moved into pending_to_remove_session_handles_, we just ingore this
+              const auto operation = (nullptr != session) ? session->GetOperation().get() : nullptr;
+
+              // If it's already moved into pending_to_remove_session_handles_, we just ignore this
               // message.
-              if (nullptr != session && session->GetOperation())
+              if (operation)
               {
                 // Session can not be destroyed when calling PerformCurlMessage
                 auto hold_session = session->shared_from_this();
-                session->GetOperation()->PerformCurlMessage(result);
+                operation->PerformCurlMessage(result);
+
+                if (operation->IsRetryable())
+                {
+                  self->pending_to_retry_sessions_.push_front(session);
```

Comment:
I believe this is safe being lock-free because it is only managed by the background thread and not public. In case the session is removed by `doAbortSessions()` or `doRemoveSessions()`, the pointer would be ignored and removed when processed by `doRetrySessions()`